### ISSUE TITLE
Polyline or Markers crashes iOS Maps (undefined is not a function)

### ISSUE THREAD

**Issue by chrisid** [react-native-maps/react-native-maps#4512]
### Summary

Trying to add any <Polyline /> or <Marker /> crashes iOS map

### Reproducible sample code

```TSX
import React from 'react';
import { Button, StyleSheet, Text } from 'react-native';

import MapView, { Marker, Polyline } from 'react-native-maps';

const ASPECT_RATIO = WINDOW_WIDTH / WINDOW_HEIGHT;
const LATITUDE_DELTA = 0.03;
const LONGITUDE_DELTA = LATITUDE_DELTA * ASPECT_RATIO;

<MapView
    pitchEnabled={false}
    rotateEnabled={false}
    zoomEnabled={false}
    scrollEnabled={false}
    showsCompass={true}
    showsUserLocation={false}
    region={{
      latitude: 51.503399,
      longitude: -0.119519,
      latitudeDelta: LATITUDE_DELTA,
      longitudeDelta: LONGITUDE_DELTA,
    }}
    onRegionChange={onRegionChange}
    style={styles.map}>
      <Marker
        key="userlocmarker"
        coordinate={{
          latitude: 51.503399,
          longitude: -0.119519,
        }}
      />
</MapView>

const styles = StyleSheet.create({
  map: {
    ...StyleSheet.absoluteFillObject,
  },
});
```

### Steps to reproduce

Add any Marker or Polyline to an iOS map

### Expected result

Render the layer

### Actual result

Undefined is not a function

![2022-11-03_11-04-38](https://user-images.githubusercontent.com/140213/199705224-8adb2f7e-120e-4b0c-b138-924b29cfe230.png)

### React Native Maps Version

1.3.2

### What platforms are you seeing the problem on?

iOS (Apple Maps)

### React Native Version

0.70.2

### What version of Expo are you using?

Not using Expo

### Device(s)

Simulator iPhone 14 Pro

**Comment by tochsner** [comment]
I had the same problem but solved at least in my case.

The issue appears when the `getNativeComponent` field of the `MapMarker` class is assigned by the decorator by setting the attribute on the prototype (`MapMarker.prototype. getNativeComponent  = ...`). If I understand it correctly, setting an attribute on a prototype only affects object instances if the field does not already exist (either if a value is assigned or if it is defined as a property) (see [here](https://stackoverflow.com/questions/22759242/override-public-field-javascript-prototype)). Otherwise, the field remains `undefined`, leading to this error when we actually want to use call `getNativeComponent`.

Now in the `MapMarker` class, the property `getNativeComponent` is already defined as a public uninitialized field. However, depending on your babel and typescript configuration, this public uninitialized field is emitted in the transpiled code, thus allowing the decorator to modify the field.

When comparing to the example project, I noticed that I had `ES2022` as module and lib configured as opposed to `ES2017`. This is relevant as with `ES2022`, uninitialized public fields are no longer emitted. Thus I had to set useDefineForClassFields to false (see [here](https://www.typescriptlang.org/tsconfig#useDefineForClassFields)).

Furthermore, I also had to adapt my babel configuration as follows:

```
presets: ['module:metro-react-native-babel-preset', '@babel/preset-typescript'],
plugins: [
    [
      'module-resolver', { ... }
    ]
]
```

I think the plugins that might cause problems are `@babel/plugin-proposal-class-properties` and `@babel/plugin-proposal-private-methods`.

Hope this helps!

**Comment by nriccar** [comment]
Any fix for this? I still need to use '@babel/plugin-proposal-private-methods'

**Comment by nriccar** [comment]
Solved with https://github.com/react-native-maps/react-native-maps/issues/4467

**Comment by SwhiteMHC** [comment]
@nriccar This is not solved. I am encountering the same issue.

**Comment by scarlac** [comment]
**My theory:**
TL;DR: TypeScript 4.4 upgrade adds support for ES2022, causing many classes to break due to [a new semi-incompatible "Class Fields" feature added to ECMAScript](https://github.com/tc39/proposals/blob/main/finished-proposals.md).

**Explanation:**
When a library or project `tsconfig.json` file's `"target"` is set to `"target": "ESNext"`, then you're saying "I'm okay with breaking changes" and "please give me all new features" or "don't transpile new features". One of these new features, Class Fields(?), that were introduced in ECMAScript went from a proposal to a full-on feature that wasn't compatible with previous transpilation behaviors.
This means that when you upgrade to TypeScript 4.4 or later, or if you change `tsconfig.json`'s "target" value to "ES2022" or later, then you are going to get these non-functional class properties that override the old behavior.  
This is the culprit, and it must be fixed by all of the library maintainers individually. React Native had to fix it internally as well. If you are consuming a library that's precompiled (ie. most libraries) you cannot fix it unless you use patch-package or similar node_modules modifications.

If you have this issue locally in YOUR code, you can fix it by changing your `tsconfig.json` file so that it targets a lower version of ECMAScript. It needs to be 2021 or lower:
```
{
  "compilerOptions": {
    ...
    "target": "ES2021",
```

**Proof of Theory:**
1. [Open this TypeScript Playground](https://www.typescriptlang.org/play/?target=8&module=199&ts=5.4.5#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBRSwRiCYQAYXRjRRNAA3ooAL7BDIKMjJBhLCoANVALPZik53x5S3oTGkvMSawUPGEAqFihKJQIDHYgkgFWO5gAFABKfJ0NiIISQAA8XHiAwknAAfABuNnBSkMamgfkUbiQFarBXcWn0ziMkSIBaQaHCKV4GWwRIstXzTXa3UnWAAQhNRtAAF4HaakBbrQk7Y6XSKEJxBDRY-H4gBZCirDL5lUlD0ZlNanV62BGk20M0lm3lwtJtVOVD4zgukrskUAeWkACtILBUAEPXJmJx9d7ffAAgQ2OhUO1+PXZc3W5xDS7giocM0PGQvIh4H0GMD2DQvm4ZjMOwDDMJA-CUuoiQ-vAegiF4ABEACiqwMBG7iCAhSTYgh4AOKgADkwgAHKQBICFBCoS5fGQ7zavwYFHKgjgKK40h4KATifjowxYPoKAIIg0ZiDwfGSF+0QiAYHDYgsTCsBwPBBCUQA)
2. Confirm that the "TS Config" target is set to ES2021 (you need to swap to ES2020 before you select ES2021 - a bug in that UI)
3. Observe that the output does NOT include the undefined class fields
4. Change the "TS Config" target to "ES2022"
5. Observe that the output now includes undefined class properties

**What to do if you are a...**
- **Library maintainer**: Read the above statement and make a decision if you want to downgrade your target to 2021 or lower, or change your original code.
- **Consumer of a library**: Link or paste the above statement to the maintainer of the library you believe is causing the issue.

**Comment by salah-ghanim** [comment]
@scarlac I've seen your pull request, I'm thinking of simply trying to move away from class definition and sticking with functional components, not sure how realistic that is, but I would like to try that before merging your pull request

**Comment by scarlac** [comment]
@salah-ghanim thanks for sharing your plan. I agree with that approach.

**Comment by chinomnsoawazie** [comment]
> I had the same problem but solved it at least in my case.
> 
> The issue appears when the `getNativeComponent` field of the `MapMarker` class is assigned by the decorator by setting the attribute on the prototype (`MapMarker.prototype.getNativeComponent = ...`). If I understand it correctly, setting an attribute on a prototype only affects object instances if the field does not already exist (either if a value is assigned or if it is defined as a property) (see [here](https://stackoverflow.com/questions/22759242/override-public-field-javascript-prototype)). Otherwise, the field remains `undefined`, leading to this error when we actually want to call `getNativeComponent`.
> 
> Now in the `MapMarker` class, the property `getNativeComponent` is already declared as a public uninitialized field. However, depending on your babel and typescript configuration, this public uninitialized field is emitted in the transpiled code, thus allowing the decorator to modify the field.
> 
> When comparing to the example project, I noticed that I had `ES2022` as module and lib configured as opposed to `ES2017`. This is relevant as with `ES2022`, uninitialized public fields are no longer emitted. Thus I had to set `useDefineForClassFields` to false in the typescript configuration (see [here](https://www.typescriptlang.org/tsconfig#useDefineForClassFields)):
> 
> ```
> {
>   ...
>   "compilerOptions": {
>     "useDefineForClassFields": false,
>     ...
>   }
>   ...
> }
> ```
> 
> Furthermore, I also had to adapt my babel configuration as follows:
> 
> ```
> presets: ['module:metro-react-native-babel-preset', '@babel/preset-typescript'],
> plugins: [
>     [
>       'module-resolver', { ... }
>     ]
> ]
> ```
> 
> I think the plugins that might cause problems are `@babel/plugin-proposal-class-properties` and `@babel/plugin-proposal-private-methods`.
> 
> Hope this helps!

Thanks so much much for this! Only addition I have is to move the babel plugins into an override this way:

```
overrides: [
    {
       test: (fileName) => !fileName.includes('node_modules/react-native-maps'),
       plugins: [
                  ['@babel/plugin-transform-private-methods', {loose: true}],
                  ['@babel/plugin-transform-class-properties', {loose: true}],
                  ['@babel/plugin-transform-private-property-in-object', {loose: true}],
              ],
      },
   ],
```

**Comment by neeraj-tangariya** [comment]
In my case scenario i did this
```
module.exports = {
  presets: ['module:@react-native/babel-preset'],
  plugins: [
    'react-native-reanimated/plugin',
    ['@babel/plugin-proposal-decorators', {legacy: true}],
    ['@babel/plugin-proposal-class-properties', {loose: true}],
    ['@babel/plugin-transform-class-properties', {loose: true}],
  ],
  overrides: [
    {
      test: fileName => !fileName.includes('node_modules/react-native-maps'),
      plugins: [
        [require('@babel/plugin-transform-private-methods'), {loose: true}],
      ],
    },
  ],
};

```
